src/main/java/uk/gov/digital/ho/hocs/audit/export/parsers/AbstractDataParser.java

```diff
@@ -0,0 +1,97 @@
+import uk.gov.digital.ho.hocs.audit.export.caseworkclient.CaseworkClient;
+import uk.gov.digital.ho.hocs.audit.export.caseworkclient.dto.GetCaseReferenceResponse;
+import uk.gov.digital.ho.hocs.audit.export.infoclient.InfoClientSupplier;
+import uk.gov.digital.ho.hocs.audit.utils.UuidStringChecker;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+public abstract class AbstractDataParser {
+
+    protected final ZonedDateTimeConverter zonedDateTimeConverter;
+    protected final ObjectMapper objectMapper;
+    protected final InfoClientSupplier infoClientSupplier;
+    protected final boolean convert;
+    private final CaseworkClient caseworkClient;
+
+    protected final Map<String, String> uuidToName;
+    protected final Map<String, String> entityListItemToName;
+
+    protected AbstractDataParser(InfoClientSupplier infoClientSupplier,
+                                 CaseworkClient caseworkClient,
+                                 ObjectMapper objectMapper,
+                                 ZonedDateTimeConverter zonedDateTimeConverter,
+                                 boolean convert) {
+        this.infoClientSupplier = infoClientSupplier;
+        this.convert = convert;
+        this.zonedDateTimeConverter = zonedDateTimeConverter;
+        this.objectMapper = objectMapper;
+        this.caseworkClient = caseworkClient;
+
+        uuidToName = new HashMap<>();
+        entityListItemToName = new HashMap<>();
+
+        if (convert) {
+            initialiseConversionValues(getUuidSuppliers(), getEntityLists());
+        }
+    }
+
+    private void initialiseConversionValues(Stream<Supplier<Map<String, String>>> uuidSuppliers,
+                                            List<String> entityLists) {
+        uuidToName.putAll(
+                uuidSuppliers
+                        .map(CompletableFuture::supplyAsync)
+                        .map(CompletableFuture::join)
+                        .collect(HashMap::new, Map::putAll, Map::putAll));
+
+        for (String listName: entityLists) {
+            entityListItemToName.putAll(infoClientSupplier.getEntityList(listName).get());
+        }
+    }
+
+
+    protected String convertValue(String value) {
+        if (!convert && value != null) {
+            return value;
+        }
+
+        if (UuidStringChecker.isUUID(value))  {
+            return uuidToName.getOrDefault(value, value);
+        }
+
+        return entityListItemToName.getOrDefault(value, value);
+    }
+
+    protected String convertCaseUuid(String value) {
+        if (!convert || !UuidStringChecker.isUUID(value)) {
+            return value;
+        }
+
+        GetCaseReferenceResponse caseReferenceResponse = caseworkClient.getCaseReference(value);
+
+        String referenceNotFound = "REFERENCE NOT FOUND";
```

Comment:
Where does the "REFERENCE NOT FOUND" string come from? If from casework, are we OK with this approach as it feels like it introduces a dependancy between audit and casework around the string value

Reply:
It comes from casework as is the current approach to tell if a case is null. I can pull this from here for now.